packages/components/src/internal/components/search/FilterFacetedSelector.tsx

```diff
@@ -1,64 +1,162 @@
+    showSearchLength?: number; // show search box if number of unique values > N
 }
 
 export const FilterFacetedSelector: FC<Props> = memo(props => {
-    const { selectDistinctOptions } = props;
+    const { api, selectDistinctOptions, fieldKey, fieldFilter, onFieldFilterUpdate, showSearchLength } = props;
 
-    const [activeFieldDistinctValues, setActiveFieldDistinctValues] = useState<any[]>(undefined);
+    const [fieldDistinctValues, setFieldDistinctValues] = useState<string[]>(undefined);
     const [error, setError] = useState<string>(undefined);
+    const [searchStr, setSearchStr] = useState<string>(undefined);
 
     useEffect(() => {
-        Query.selectDistinctRows({
-            ...selectDistinctOptions,
-            success: result => {
-                const distinctValues = result.values.sort(naturalSort);
-                setActiveFieldDistinctValues(distinctValues);
-            },
-            failure: error => {
+        api.query
+            .selectDistinctRows(selectDistinctOptions)
+            .then(result => {
+                const distinctValues = result.values
+                    .map(val => {
+                        if (val === '' || val === null || val === undefined) return EMPTY_VALUE_DISPLAY;
+                        return val;
+                    })
+                    .sort(naturalSort);
+                distinctValues.unshift(ALL_VALUE_DISPLAY);
+                setFieldDistinctValues(distinctValues);
+            })
+            .catch(error => {
                 console.error(error);
                 setError(resolveErrorMessage(error));
-            },
+            });
+    }, [selectDistinctOptions, fieldKey]);
+
+    const checkedValues = useMemo(() => {
+        return getCheckedFilterValues(fieldFilter, fieldDistinctValues);
+    }, [fieldFilter, fieldKey, fieldDistinctValues]); // need to add fieldKey
+
+    const taggedValues = useMemo(() => {
+        if (checkedValues?.indexOf(ALL_VALUE_DISPLAY) > -1) return [];
+        return checkedValues;
+    }, [checkedValues]);
+
+    const onSearchStrChange = useCallback(e => {
+        setSearchStr(e.target.value);
+    }, []);
+
+    const onChange = useCallback(
+        (value: string, checked: boolean, uncheckOthers?: boolean) => {
+            const newFilter = getUpdatedChooseValuesFilter(
+                fieldDistinctValues,
+                fieldKey,
+                value,
+                checked,
+                fieldFilter,
+                uncheckOthers
+            );
+            onFieldFilterUpdate(newFilter);
+        },
+        [fieldDistinctValues, fieldKey, fieldFilter, onFieldFilterUpdate]
+    );
+
+    const filteredFieldDistinctValues = useMemo(() => {
+        if (!searchStr) return fieldDistinctValues;
+
+        return fieldDistinctValues?.filter(val => {
+            let valuesToKeep = [];
+            if (checkedValues?.indexOf(ALL_VALUE_DISPLAY) === -1) valuesToKeep = checkedValues;
```

Comment:
this `.filter()` function seems odd to me. Should the setting of the `valuesToKeep` be done outside of the filter function since it has nothing to do with the `val`?